src/reactive-membrane.ts

```diff
@@ -89,10 +84,10 @@ export class ReactiveMembrane {
         const unwrappedValue = unwrap(value);
         const distorted = this.valueDistortion(unwrappedValue);
         if (this.valueIsObservable(distorted)) {
-            const o = this.getReactiveState(unwrappedValue, distorted);
+            const readOnly = this.getReactiveHandler(unwrappedValue, distorted, true);
             // when trying to extract the writable version of a readonly
             // we return the readonly.
-            return o.readOnly === value ? value : o.reactive;
+            return readOnly === value ? value : this.getReactiveHandler(unwrappedValue, distorted, false);
```

Comment:
Yeah, this code path is only hit for [`@track` fields](https://github.com/salesforce/lwc/blob/d93e2963d8d32b138ce9ed130d98e7bef84b3d74/packages/%40lwc/engine-core/src/framework/decorators/track.ts#L27), and we do not have any perf test with `@track`.

@nolanlawson This makes me wonder if we could remove the creation of the readOnly proxy in this routine? my mental model is: this check is in place for when we want to `getProxy` of a readonly proxy, therefore it should already be in the `readOnlyObjectGraph`, and if the value is not in the `readOnlyObjectGraph`, then we should create a `this.getReactiveHandler(unwrappedValue, distorted, false)`.

Even if is not visible in the perf tests (because all of their values are `@api`/readonly), for cases in which the values are set to a writable proxy (ex: data from lds) or `@track foo = {...}`, it will be noticeable, since we won't create a read-only proxy when we shouldn't (mem + processing time)

Reply:
Thanks for the details! I'll add a benchmark for `@track`. I think it's important to test this use case.

As for not creating the extra `readOnly`, your explanation makes sense to me. We should be able to validate it with a `@track` benchmark.

Reply:
@jodarove Made the change you recommended, and added a couple tests for safety. I'll re-run the perf tests with `@track` included.